pkg/common-controller/snapshot_controller.go

```diff
@@ -337,16 +337,10 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 		}
 
 		// update snapshot status
-		for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
-			klog.V(5).Infof("syncUnreadySnapshot [%s]: trying to update snapshot status", utils.SnapshotKey(snapshot))
-			_, err = ctrl.updateSnapshotStatus(snapshot, newContent)
-			if err == nil {
-				break
-			}
-			klog.V(4).Infof("failed to update snapshot %s status: %v", utils.SnapshotKey(snapshot), err)
-		}
-
+		klog.V(5).Infof("syncUnreadySnapshot [%s]: trying to update snapshot status", utils.SnapshotKey(snapshot))
+		_, err = ctrl.updateSnapshotStatus(snapshot, newContent)
 		if err != nil {
+			klog.V(4).Infof("failed to update snapshot %s status: %v", utils.SnapshotKey(snapshot), err)
 			// update snapshot status failed
 			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotStatusUpdateFailed", fmt.Sprintf("Snapshot status update failed, %v", err))
 			return err
```

Comment:
Looking at the caller, I see that we don't requeue the object and instead rely on resync period to retry. I think we should change the behavior to requeue, but we should also make sure the workqueue has proper throttling/backoff as well.

Reply:
Use queue.AddRateLimited(key)?

Reply:
Can you take a look of this? https://github.com/kubernetes-csi/external-snapshotter/pull/211/commits/10017c2bdccc95268596aefdcbdc769bc5b2dc3c

Reply:
Yes, I like this direction. Also cc @yuxiangqian and @jsafrane for more eyes.  This will cause snapshotter to potentially retry more frequently than it has in the past, although it should be rate limited.

Reply:
I wonder if we should delay this change.  This changes snapshot controller behavior a lot.  In the past we don't do retries when create snapshot fails because snapshot is time sensitive (other than a few retries when creating/updating API objects). 

Can we revisit this after the 2.0 release?  We have a pending fix for the create snapshot timeout issue as well that we should go back to after 2.0.

Reply:
Ok let's open an issue to track it.

Reply:
Sure

Reply:
Opened an issue.  Let me remove this re-queue logic from this PR.

Reply:
A little concerned about removing retrying logic for status update and snapshot content creation from application perspective. Do we have a rough idea how often retry would resolve API server failures? Application could end up based of the existence of status field of a snapshot to decide whether or not to unquiesce. With RateLimited requeue, the waiting time could be significantly longer?

Reply:
Removed snapshotQueue.AddRateLimited().

What is conclusion here?  Should I just add back createSnapshotContentRetryCount() as a internal const or is the current PR okay?  @msau42

Reply:
I don't like having retries within retries. It makes timing hard to reason about (metrics look strange), and we can end up overloading the API server since there's two different throttling behaviors that need to be controlled.  We previously had issues in external-provisioner due to it having multiple retries within retries and I'd like to avoid that here too. We can tune the rate limiting as necessary.

But I agree this is a large, fundamental change to how the snapshot controller has operated, so I'm ok deferring it to the next release.

Reply:
Yes, let's add the retry count back for now so the only change is not making the retry count configurable.

Reply:
Sure.